r2-core/src/main/java/com/linkedin/r2/transport/http/client/AsyncPoolImpl.java

```diff
@@ -255,13 +278,15 @@ public void start()
         throw new IllegalStateException(_poolName + " is " + _state);
       }
       _state = State.RUNNING;
-      if (_idleTimeout > 0)
+      long timeOut = Math.min(_idleTimeout, _waiterTimeout);
+      if (timeOut > 0)
       {
-        long freq = Math.min(_idleTimeout / 10, 1000);
+        long freq = Math.max(MIN_WAITER_TIMEOUT, Math.min(timeOut / 10, MAX_WAITER_TIMEOUT));
```

Comment:
Thinking more about this, instead of modifying the min timeout (which could have an impact on servers with thousand of AsyncPools when set to a low value), we could run the expire logic per request.

This would guarantee us that we are doing the checking work only when actually needed and not keep checking every 100-300ms x 5000 Async Pools which can be a lot of load on the servers if we don't actually have any new connection.

Reply:
Thanks for the offline discussion. This absolutely make sense. I have pushed the changes as discussed.